snapshots/devmapper/pool_device.go

```diff
@@ -436,6 +437,16 @@ func (p *PoolDevice) DeactivateDevice(ctx context.Context, deviceName string, de
 
 	if err := p.transition(ctx, deviceName, Deactivating, Deactivated, func() error {
 		return retry(ctx, func() error {
+			// device nodes only exist for non-committed snapshots
+			device, err := os.OpenFile(deviceName, os.O_RDWR, 0)
+			if err != nil && !os.IsNotExist(err) {
+				log.G(ctx).Warnf("unable to open device %q to flush buffers: %+v", deviceName, err)
+			} else if err == nil {
+				defer device.Close()
+				if err := unix.IoctlSetInt(int(device.Fd()), unix.BLKFLSBUF, 0); err != nil {
+					log.G(ctx).Warnf("could not flush buffers for device %q: %+v", deviceName, err)
+				}
+			}
```

Comment:
Some thoughts;

- It looks like the `device` is only needed within this block; perhaps would be good to make that explicit, make the variable local, and put everything in a conditional if/else block.
- I'm always a bit scary of `if foo && bar .. else if` conditions; combining multiple conditions can be error-prone, and it's easy to overlook "other" conditions (and miss them)
- As `device` is only used in this block, perhaps the `device.Close()` should not be done in a `defer`, but immediately after we're done?
- ^^ if it's not closed immediately, could it cause the `dmsetup.DiscardBlocks()` to return a `dmsetup.ErrInUse` ?


Perhaps something like this would address those points above;

 ```suggestion
			// device nodes only exist for non-committed snapshots
			if device, err := os.OpenFile(deviceName, os.O_RDWR, 0); err != nil {
				if !os.IsNotExist(err) {
					log.G(ctx).Warnf("unable to open device %q to flush buffers: %+v", deviceName, err)
				}
			} else {
				if err := unix.IoctlSetInt(int(device.Fd()), unix.BLKFLSBUF, 0); err != nil {
					log.G(ctx).Warnf("could not flush buffers for device %q: %+v", deviceName, err)
				}
				_ = device.Close()
			}
```

As the `deviceName` is also used further down, also wondering if errors from `device.Close()` should be fully ignored, or (at least?) logged.

Reply:
I think moving this to a separate function would ameliorate the control flow bug-bait, plus allow for flushing before and after `DiscardBlocks`.

For `close()`, I think logging is probably a good call-out, looking at the potential error cases in the man page, those look pretty gnarly. I think the Go runtime will auto-retry on EINTR, EIO may not apply since the buffers were just flushed but an iop could sneak in. If the devmapper pool is backed by files on NFS, I think the admin deserves those nasal demons 😅